// firebase-tools/lib/gcp/docker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = void 0;
const error_1 = require("../error");
const api = require("../apiv2");
function isErrors(response) {
    return Object.prototype.hasOwnProperty.call(response, "errors");
}
const API_VERSION = "v2";
class Client {
    constructor(origin) {
        this.client = new api.Client({
            apiVersion: API_VERSION,
            auth: true,
            urlPrefix: origin,
        });
    }
    async listTags(path) {
        const response = await this.client.get(`${path}/tags/list`);
        if (isErrors(response.body)) {
            throw new error_1.FirebaseError(`Failed to list GCR tags at ${path}`, {
                children: response.body.errors,
            });
        }
        return response.body;
    }
    async deleteTag(path, tag) {
        var _a;
        const response = await this.client.delete(`${path}/manifests/${tag}`);
        if (((_a = response.body.errors) === null || _a === void 0 ? void 0 : _a.length) != 0) {
            throw new error_1.FirebaseError(`Failed to delete tag ${tag} at path ${path}`, {
                children: response.body.errors,
            });
        }
    }
    async deleteImage(path, digest) {
        var _a;
        const response = await this.client.delete(`${path}/manifests/${digest}`);
        if (((_a = response.body.errors) === null || _a === void 0 ? void 0 : _a.length) != 0) {
            throw new error_1.FirebaseError(`Failed to delete image ${digest} at path ${path}`, {
                children: response.body.errors,
            });
        }
    }
}
exports.Client = Client;
