// firebase-tools/lib/deploy/functions/errorHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorHandler = void 0;
const clc = require("cli-color");
const logger_1 = require("../../logger");
const functionsDeployHelper_1 = require("./functionsDeployHelper");
const error_1 = require("../../error");
class ErrorHandler {
    constructor() {
        this.errors = [];
        this.warnings = [];
    }
    record(level, functionName, operationType, message) {
        const info = {
            functionName,
            operationType,
            message,
        };
        if (level === "error") {
            this.errors.push(info);
        }
        else if (level === "warning") {
            this.warnings.push(info);
        }
    }
    printErrors() {
        if (this.errors.length === 0) {
            return;
        }
        logger_1.logger.info("");
        logger_1.logger.info("Functions deploy had errors with the following functions:");
        for (const failedDeployment of this.errors) {
            logger_1.logger.info(`\t${functionsDeployHelper_1.getFunctionLabel(failedDeployment.functionName)}`);
        }
        const failedIamCalls = this.errors.filter((e) => e.operationType === "set invoker");
        if (failedIamCalls.length) {
            logger_1.logger.info("");
            logger_1.logger.info("Unable to set the invoker for the IAM policy on the following functions:");
            for (const failedDep of failedIamCalls) {
                logger_1.logger.info(`\t${failedDep.functionName}`);
            }
            logger_1.logger.info("");
            logger_1.logger.info("Some common causes of this:");
            logger_1.logger.info("");
            logger_1.logger.info("- You may not have the roles/functions.admin IAM role. Note that roles/functions.developer does not allow you to change IAM policies.");
            logger_1.logger.info("");
            logger_1.logger.info("- An organization policy that restricts Network Access on your project.");
        }
        logger_1.logger.info("");
        logger_1.logger.info("To try redeploying those functions, run:");
        logger_1.logger.info("    " +
            clc.bold("firebase deploy --only ") +
            clc.bold('"') +
            clc.bold(this.errors
                .map((failedDeployment) => `functions:${functionsDeployHelper_1.getFunctionId(failedDeployment.functionName).replace(/-/g, ".")}`)
                .join(",")) +
            clc.bold('"'));
        logger_1.logger.info("");
        logger_1.logger.info("To continue deploying other features (such as database), run:");
        logger_1.logger.info("    " + clc.bold("firebase deploy --except functions"));
        for (const failedDeployment of this.errors) {
            logger_1.logger.debug(`\tError during ${failedDeployment.operationType} for ${failedDeployment.functionName}: ${failedDeployment.message}`);
        }
        throw new error_1.FirebaseError("Functions did not deploy properly.");
    }
    printWarnings() {
        if (this.warnings.length === 0) {
            return;
        }
        for (const failedDeployment of this.warnings) {
            logger_1.logger.debug(`\tWarning during${failedDeployment.operationType} for ${failedDeployment.functionName}: ${failedDeployment.message}`);
        }
    }
}
exports.ErrorHandler = ErrorHandler;
